src/main/scala/org/ergoplatform/nodeView/mempool/OrderedTxPool.scala

```diff
@@ -32,48 +36,100 @@ final case class OrderedTxPool(orderedTransactions: TreeMap[WeightedTxId, ErgoTr
   def size: Int = orderedTransactions.size
 
   def get(id: ModifierId): Option[ErgoTransaction] = {
-    transactionsRegistry.get(id).flatMap(w => orderedTransactions.get(WeightedTxId(id, w)))
+    transactionsRegistry.get(id).flatMap(orderedTransactions.get(_))
   }
 
+
+  /**
+    * Add new transaction to the pool and throw away from the pool transaction with the smallest weight
+    * if pool is overflown. We should first add transaction and only after it find candidate for replacement
+    * because new transaction may affect weights of existed transaction in mempool (see updateFamily).
+    * So candidate for replacement (transaction with minimal weight) can be changed after adding new transaction.
+    * put() is preceded by canAccept method which enforces that newly added transaction will not be immediately
+    * thrown from the pool.
+    *
+    * @param tx - transaction to add
+    * @return - modified pool
+    */
   def put(tx: ErgoTransaction): OrderedTxPool = {
-    val (txs, ws) = if (orderedTransactions.size >= mempoolCapacity) {
-      val keyToThrow = orderedTransactions.firstKey
-      log.debug(s"Memory pool overflow - throwing out transaction with id = ${keyToThrow.id}")
-      (orderedTransactions - keyToThrow) -> (transactionsRegistry - keyToThrow.id)
+    val wtx = weighted(tx)
+    val newPool = OrderedTxPool(orderedTransactions.updated(wtx, tx),
+      transactionsRegistry.updated(wtx.id, wtx), invalidated,
+      outputs ++ tx.outputs.map(_.id -> wtx)).updateFamily(tx, wtx.weight)
+    if (newPool.orderedTransactions.size > mempoolCapacity) {
+      val victim = newPool.orderedTransactions.head._2
+      newPool.remove(victim)
     } else {
-      orderedTransactions -> transactionsRegistry
+      newPool
     }
-    val wtx = weighted(tx)
-    OrderedTxPool(txs.updated(wtx, tx), ws.updated(wtx.id, wtx.weight), invalidated)
   }
 
   def remove(tx: ErgoTransaction): OrderedTxPool = {
-    OrderedTxPool(orderedTransactions - weighted(tx), transactionsRegistry - tx.id, invalidated)
+    transactionsRegistry.get(tx.id).fold(this)(wtx =>
+      OrderedTxPool(orderedTransactions - wtx, transactionsRegistry - tx.id, invalidated, outputs -- tx.outputs.map(_.id)).updateFamily(tx, -wtx.weight))
   }
 
   def invalidate(tx: ErgoTransaction): OrderedTxPool = {
     val inv = if (invalidated.size >= blacklistCapacity) invalidated - invalidated.firstKey else invalidated
     val ts = System.currentTimeMillis()
-    OrderedTxPool(orderedTransactions - weighted(tx),
-      transactionsRegistry - tx.id, inv.updated(tx.id, ts))
+    transactionsRegistry.get(tx.id).fold(OrderedTxPool(orderedTransactions, transactionsRegistry, inv.updated(tx.id, ts), outputs))(wtx =>
+      OrderedTxPool(orderedTransactions - wtx, transactionsRegistry - tx.id, inv.updated(tx.id, ts),
+        outputs -- tx.outputs.map(_.id)).updateFamily(tx, -wtx.weight))
```

Comment:
Lets avoid complex expressions inside function calls, it makes code less readable

Reply:
Or at least use named arguments

Reply:
Sorry, I do not have much experience in Scala programming, so I don't know how to satisfy your request. Certainly I can rewrite this code in C-like way:

```
val entry = transactionsRegistry.get(tx.id)
if (entry.isEmpty) {
   OrderedTxPool(orderedTransactions, transactionsRegistry, inv.updated(tx.id, ts), outputs)
} else {
   val wtx =  entry.get
   val newPool = OrderedTxPool(orderedTransactions - wtx, 
       transactionsRegistry - tx.id, 
       inv.updated(tx.id, ts),
       outputs -- tx.outputs.map(_.id))
   newPool.updateFamily(tx, -wtx.weight)
}
```

I have used fold() thinking that it is more functional and "Scala" style.

Your suggestion to use named arguments seems to have no sense here, because names of parameters are the same as names of arguments. So writing `OrderedTxPool(orderedTransactions=orderedTransactions,...)`
just more complicates the expression and even making it more obscure.